### PR TITLE
feat(cmd-api-server): swagger.json endpoints

### PR DESCRIPTION
**Pull request by charellesandig**
commit ff29e94e981a80a8665d82f2190f34f79087531e (HEAD -> pr/charellesandig/1994, charellesandig/issue1877)
Author: Peter Somogyvari <peter.somogyvari@accenture.com>
Date:   Wed Sep 13 16:39:42 2023 -0700

    feat(cmd-api-server): add GetOpenApiSpecV1Endpoint (HTTP GET)

    The new endpoint can serve up the OpenAPI specification file of the
    API server via a get request without any parameters needed.

    Under the hood it uses the re-usable common base endpoint class that
    we've recently added to the core package.

    The test case at this file path demonstrates how to use it via the API
    client:
    `packages/cactus-cmd-api-server/src/test/typescript/unit/get-open-api-spec-v1-endpoint.test.ts`

    Signed-off-by: Peter Somogyvari <peter.somogyvari@accenture.com>

commit ebd0abfebdfd6051ca61d278f16b49787c168887
Author: Peter Somogyvari <peter.somogyvari@accenture.com>
Date:   Wed Sep 13 14:09:30 2023 -0700

    feat(connector-besu): add GetOpenApiSpecV1Endpoint (HTTP GET)

    The new endpoint can serve up the OpenAPI specification file of the
    plugin via a get request without any parameters needed.

    Under the hood it uses the re-usable common base endpoint class that
    we've recently added to the core package.

    The test case at this file path demonstrates how to use it via the API
    client:
    `packages/cactus-plugin-ledger-connector-besu/src/test/typescript/
    unit/get-open-api-spec-v1-connector-besu.test.ts`

    Signed-off-by: Peter Somogyvari <peter.somogyvari@accenture.com>

commit 4b541ec8bf30bd5d0df7feb6caefc1c1b0c21c12
Author: Peter Somogyvari <peter.somogyvari@accenture.com>
Date:   Wed Sep 13 13:48:20 2023 -0700

    feat(cactus-core): add GetOpenApiSpecV1EndpointBase<S, P> class

    This is the pre-requisite to finishing the task at
    https://github.com/hyperledger/cacti/issues/1877

    Having this generic endpoint class available will allow us to send in
    another commit which then uses it to create the Open API spec endpoints
    in all the plugin classes.

    Example usage of the generic class looks like this:

    ```typescript
    import {
      GetOpenApiSpecV1EndpointBase,
      IGetOpenApiSpecV1EndpointBaseOptions,
    } from "@hyperledger/cactus-core";

    import { Checks, LogLevelDesc } from "@hyperledger/cactus-common";
    import { IWebServiceEndpoint } from "@hyperledger/cactus-core-api";

    import OAS from "../../json/openapi.json";

    export const OasPathGetOpenApiSpecV1 =
      OAS.paths[
        "/api/v1/plugins/@hyperledger/cactus-plugin-ledger-connector-besu/get-open-api-spec"
      ];

    export type OasPathTypeGetOpenApiSpecV1 = typeof OasPathGetOpenApiSpecV1;

    export interface IGetOpenApiSpecV1EndpointOptions
      extends IGetOpenApiSpecV1EndpointBaseOptions<
        typeof OAS,
        OasPathTypeGetOpenApiSpecV1
      > {
      readonly logLevel?: LogLevelDesc;
    }

    export class GetOpenApiSpecV1Endpoint
      extends GetOpenApiSpecV1EndpointBase<typeof OAS, OasPathTypeGetOpenApiSpecV1>
      implements IWebServiceEndpoint
    {
      public get className(): string {
        return GetOpenApiSpecV1Endpoint.CLASS_NAME;
      }

      constructor(public readonly options: IGetOpenApiSpecV1EndpointOptions) {
        super(options);
        const fnTag = `${this.className}#constructor()`;
        Checks.truthy(options, `${fnTag} arg options`);
      }
    }
    ```

    And the associated OpenAPI specification's paths entry:

    ```json
    "/api/v1/plugins/@hyperledger/cactus-plugin-ledger-connector-besu/get-open-api-spec": {
      "get": {
        "x-hyperledger-cactus": {
          "http": {
            "verbLowerCase": "get",
            "path": "/api/v1/plugins/@hyperledger/cactus-plugin-ledger-connector-besu/get-open-api-spec"
          }
        },
        "operationId": "getOpenApiSpecV1",
        "summary": "Retrieves the .json file that contains the OpenAPI specification for the plugin.",
        "parameters": [],
        "responses": {
          "200": {
            "description": "OK",
            "content": {
              "application/json": {
                "schema": {
                  "type": "string"
                }
              }
            }
          }
        }
      }
    },
    ```

    Signed-off-by: Peter Somogyvari <peter.somogyvari@accenture.com>

### CODE REVIEW

**Review by takeutak — APPROVED** [review]

File: packages/cactus-core/src/main/typescript/web-services/swagger-endpoints.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import { OpenApiJson } from "@hyperledger/cactus-core-api";
+import express from "express";
+const app = express();
+import swaggerUi from "swagger-ui-express";
+
+const options = {
```
Comment (petermetz) [inline packages/cactus-core/src/main/typescript/web-services/swagger-endpoints.ts:6]: @charellesandig We cannot hardcode the URLs/names, they must come from the  web service plugins dynamically at runtime. Please take a look at the `PluginRegistry` class and how it is used by the `ApiServer` to do such enumerations and then work your way from there. I'm happy to assist with more code references.

**Review by petermetz — CHANGES_REQUESTED** [review]

**Review by jagpreetsinghsasan — DISMISSED** [review]
Just one question here. @petermetz  do we also need to have to patch our test cases with relevant code, to include testing of this endpoint as well?

File: extensions/cactus-plugin-htlc-coordinator-besu/src/main/typescript/web-services/get-openapi-spec.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  registerWebServiceEndpoint,
+} from "@hyperledger/cactus-core";
+import OAS from "../../json/openapi.json";
+
+export interface IGetOpenApiSpecEndpointOptions {
+  logLevel?: LogLevelDesc;
+  pluginRegistry: PluginRegistry;
+}
+
+export class GetOpenApiSpecEndpoint implements IWebServiceEndpoint {
```
Comment (petermetz) [inline extensions/cactus-plugin-htlc-coordinator-besu/src/main/typescript/web-services/get-openapi-spec.ts:27]: @charellesandig Please create one endpoint class that is generic across the board and can be re-used by all the plugins when they are registering their web-services.
The `OAS` should be injected through a constructor argument which the plugins have the ability to provide.
This way you don't need to copy paste the same endpoint code to literally every plugin we have so far.

Please also make sure to update the `CLASS_NAME` variable because right now it's set to some other value.

Please also make sure to match the file name to the class name (right now the word endpoint is missing).

Also please add v1 to the classname and the file name as well so classname could be `GetOpenApiSpecV1Endpoint` and the filename `get-open-api-spec-v1-endpoint.ts`

I'm copy pasting here the relevant acceptance criteria that I'm talking about for re-usability:

> The code for this web service should be somehow reused across plugins instead of being copy pasted over to each and every one of them. Probably the cactus-core package is a good place to start, but this might not be the best solution, make sure to check.

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-htlc-eth-besu/src/main/typescript/web-services/get-swagger-endpoint.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+} from "@hyperledger/cactus-common";
+import {
+  IEndpointAuthzOptions,
+  IExpressRequestHandler,
+  IWebServiceEndpoint,
+} from "@hyperledger/cactus-core-api";
+import { registerWebServiceEndpoint } from "@hyperledger/cactus-core";
+import OAS from "../../json/openapi.json";
+
+//import { InitializeRequest } from "../generated/openapi/typescript-axios";
```
[the remark this answers is not in the document]
Comment (petermetz) [reply packages/cactus-plugin-htlc-eth-besu/src/main/typescript/web-services/get-swagger-endpoint.ts]: > Can this code be removed, as its not required?

+1 Please remove code that is not being used (e.g. commented and then forgotten like that)
@charellesandig

**Comment by petermetz** [comment]
> Just one question here. @petermetz do we also need to have to patch our test cases with relevant code, to include testing of this endpoint as well?

@jagpreetsinghsasan Yes, there will have to be test coverage for these endpoints as well but I don't want to get into that yet because we are still working out the fundamental acceptance criteria such as re-usability (see my submitted change request above for details).

File: packages/cactus-plugin-htlc-eth-besu/src/main/typescript/web-services/get-swagger-endpoint.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (charellesandig) [reply packages/cactus-plugin-htlc-eth-besu/src/main/typescript/web-services/get-swagger-endpoint.ts]: Working on it.

File: extensions/cactus-plugin-htlc-coordinator-besu/src/main/typescript/web-services/get-openapi-spec.ts (hunk 2 above)
Comment (charellesandig) [reply extensions/cactus-plugin-htlc-coordinator-besu/src/main/typescript/web-services/get-openapi-spec.ts]: Working on it Peter. Thank you.

File: packages/cactus-cmd-api-server/tsconfig.json [hunk 4]
```diff
@@ -11,7 +11,7 @@
   "include": [
     "./src",
     "src/**/*.json"
-  ],
+, "src/main/typescript/openapi/.ts"  ],
```
Comment (petermetz) [inline packages/cactus-cmd-api-server/tsconfig.json:14]: @charellesandig Why was this addition necessary? Don't the existing include patterns cover this path already?

File: packages/cactus-cmd-api-server/src/main/json/openapi.json [hunk 5]
```diff
@@ -74,6 +74,10 @@
             "PrometheusExporterMetricsResponse": {
                 "type": "string",
                 "nullable": false
+            },
+            "GetOpenApiSpecResponse": {
```
Comment (petermetz) [inline packages/cactus-cmd-api-server/src/main/json/openapi.json:78]: @charellesandig Please include the major  version number in the model name so that once we do more major releases we can more easily maintain the different versions in parallel. Please see the other model names for the existing naming convention to mimic. (the ones where the convention does exist - unfortunately not all models)

File: packages/cactus-cmd-api-server/src/main/json/openapi.json [hunk 6]
```diff
@@ -128,6 +132,31 @@
...
+          "/api/v1/api-server/get-open-api-spec-v1-endpoint": {
+            "post": {
+                 "description": "Returns the openapi.json document of specific plugin.",
+                 "x-hyperledger-cactus": {
+                     "http": {
+                         "verbLowerCase": "get",
+                         "path": "/api/v1/api-server/get-open-api-spec-v1-endpoint"
+                     }
+                 },
+                  "operationId": "getOpenApiSpec",
```
Comment (petermetz) [inline packages/cactus-cmd-api-server/src/main/json/openapi.json:145]: @charellesandig Please include the major  version number in the name so that once we do more major releases we can more easily maintain the different versions in parallel. Please see the other names for the existing naming convention to mimic. (the ones where the convention does exist - unfortunately not all names)

File: packages/cactus-cmd-api-server/src/main/typescript/api-server.ts [hunk 7]
```diff
@@ -58,6 +58,7 @@ import { WatchHealthcheckV1Endpoint } from "./web-services/watch-healthcheck-v1-
 import * as default_service from "./generated/proto/protoc-gen-ts/services/default_service";
 import { GrpcServerApiServer } from "./web-services/grpc/grpc-server-api-server";
 import { determineAddressFamily } from "./common/determine-address-family";
+//import { LogLevelDesc } from "@hyperledger/cactus-common";
```
Comment (petermetz) [inline packages/cactus-cmd-api-server/src/main/typescript/api-server.ts:61]: @charellesandig Please clean up commented code from the diff

File: packages/cactus-cmd-api-server/src/main/typescript/api-server.ts [hunk 8]
```diff
@@ -184,6 +185,10 @@ export class ApiServer {
     this.log.debug("pluginsPath: %o", pluginsPath);
   }
 
+  // public getOpenApiSpec(): unknown {
```
Comment (petermetz) [inline packages/cactus-cmd-api-server/src/main/typescript/api-server.ts:188]: @charellesandig Please clean up commented code from the diff

File: packages/cactus-cmd-api-server/src/main/typescript/openapi/open-api-spec.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import { Checks } from "@hyperledger/cactus-common";
+import { PluginRegistry } from "@hyperledger/cactus-core";
+import {
+  ICactusPlugin,
+  IPluginWebService,
+  isIPluginWebService,
+} from "@hyperledger/cactus-core-api";
+import type { OpenAPIV3 } from "express-openapi-validator/dist/framework/types";
+
+export async function openApiSpec(
```
Comment (petermetz) [inline packages/cactus-cmd-api-server/src/main/typescript/openapi/open-api-spec.ts:10]: @charellesandig Please name this (and the file) as a verb not a noun.

**Review by petermetz — CHANGES_REQUESTED** [review]
@charellesandig Please write a test case that shows and verifies the endpoint working correctly at runtime.

**Review by petermetz — APPROVED** [review]
LGTM - I just did a complete refactor so now it's closer to the original acceptance criteria but it still doesn't cover all the plugins just the API server and the Besu connector. This is enough to establish a pattern based on which we can create smaller future PRs that make it happen for the rest of the plugins that have web services and Open API specs of their own.

**Comment by petermetz** [comment]
Note to self: These are the crash logs for the two jobs that are timing out for some reason. It's likely a bug with the plugin registry initialization logic of the API server that was changed a little bit to make it more flexible (more flexible = more bug prone as usual...)

[2023-09-14T20-29-18-cacti-ci-crash-2.log](https://github.com/hyperledger/cacti/files/12615099/2023-09-14T20-29-18-cacti-ci-crash-2.log)

[2023-09-14T20-29-18-cacti-ci-crash.log](https://github.com/hyperledger/cacti/files/12615100/2023-09-14T20-29-18-cacti-ci-crash.log)
